### PR TITLE
[pt] rb pruning

### PR DESCRIPTION
**Pull request by AlexanderDokuchaev**
### Changes

Add rb pruning algorithm
Rename prune folder to pruning 
Add parameters to select pruning mode in example 

### Related tickets

173791

### CODE REVIEW

File: examples/pruning/torch/resnet18/README.md [hunk 1]
```diff
@@ -32,4 +32,6 @@ It's pretty simple. The example does not require additional preparation. It will
 
 ```bash
 python main.py
+# Or to run Regularization-Based pruning
+python main.py --mode rb
```
Comment (ljaljushkin) [inline examples/pruning/torch/resnet18/README.md:36]: the [documentation about sparsity](https://github.com/openvinotoolkit/nncf/blob/develop/docs/usage/training_time_compression/other_algorithms/Sparsity.md
) is not connected with this sample and contains old configs.
are you going to update in the next PRs?

File: src/nncf/torch/function_hook/pruning/rb/modules.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+    if training:
+        uniform = torch.empty_like(mask, requires_grad=False).uniform_()
+        mask = mask + logit(uniform.clamp(eps, 1 - eps))
+
+    return input_ * binary_mask(mask)
+
+
+@COMPRESSION_MODULES.register()
+class RBPruningMask(nn.Module, StatefulModuleInterface):
```
Comment (ljaljushkin) [inline src/nncf/torch/function_hook/pruning/rb/modules.py:63]: is it possible to save and load pruned model? Is there test for such functionality?

File: tests/torch2/function_hook/pruning/magnitude/test_schedulers.py [hunk 3]
```diff
@@ -9,45 +9,54 @@
...
+
+    mask = get_pruned_modules(model)["conv.weight"].binary_mask
+    r = float(1 - mask.sum() / 81)
+    assert r == 0.5061728358268738
+    scheduler.step()
+    r = float(1 - mask.sum() / 81)
+    assert r == 0.1111111044883728
+    scheduler.step()
+    r = float(1 - mask.sum() / 81)
+    assert r == 0.48148149251937866
```
Comment (ljaljushkin) [inline tests/torch2/function_hook/pruning/magnitude/test_schedulers.py:62]: shouldn't the ratio monotonously grow?

**Comment by ljaljushkin** [comment]
Overall looks good and functional. 

I've run the sample and checked actual sparsity in IR .bin file. It's around 68%
image

```
Current pruning ratio: 0.689
Accuracy@1 of pruned model after 29 epoch: 54.870
```

**Comment by ljaljushkin** [comment]
sample has some unused imports. formatting of samples is not part of make-precommit routine, is it?

File: examples/pruning/torch/resnet18/README.md (hunk 1 above)
Comment (AlexanderDokuchaev) [reply examples/pruning/torch/resnet18/README.md:36]: yes,

File: tests/torch2/function_hook/pruning/magnitude/test_schedulers.py (hunk 3 above)
Comment (AlexanderDokuchaev) [reply tests/torch2/function_hook/pruning/magnitude/test_schedulers.py:62]: No, it's exponential scheduler

**Comment by AlexanderDokuchaev** [comment]
> sample has some unused imports. formatting of samples is not part of make-precommit routine, is it?

no, it's works
technically imports used, but it's not need, fixed

File: src/nncf/torch/function_hook/pruning/rb/modules.py (hunk 2 above)
Comment (AlexanderDokuchaev) [reply src/nncf/torch/function_hook/pruning/rb/modules.py:63]: yes, added tests

File: tests/torch2/function_hook/pruning/magnitude/test_schedulers.py (hunk 3 above)
Comment (ljaljushkin) [reply tests/torch2/function_hook/pruning/magnitude/test_schedulers.py:62]: discussed offline, the first value is not defined by scheduler, it's the initial rate. 
starting from second value it grows monotonously as expected

File: tests/torch2/function_hook/pruning/magnitude/test_algo.py [hunk 4]
```diff
@@ -113,3 +116,32 @@ def test_multi_device_infer():
...
+        "nncf_config": nncf.torch.get_config(pruned_model),
+    }
+    path_to_checkpoint = tmpdir / "checkpoint.pth"
+    torch.save(checkpoint, path_to_checkpoint)
+
+    resuming_checkpoint = torch.load(path_to_checkpoint)
+    nncf_config = resuming_checkpoint["nncf_config"]
+    state_dict = resuming_checkpoint["state_dict"]
+
+    loaded_model = ConvModel()
```
Comment (ljaljushkin) [inline tests/torch2/function_hook/pruning/magnitude/test_algo.py:141]: I'd run model and compare output before save to make sure it's loaded correctly. Should be quick for 1layer model
